File: creational_patterns/go/abstract_factory/computer.go
/**
 * @file operation.go
 * @author tushanFox (tushanFox@outlook.com)
 * @brief 抽象工厂模式，以“电脑”举例。
 * @version 0.1
 * @date 2022-02-13
 *
 * @copyright Copyright (c) 2022
 *
 */
package main

// --------------------------------------------------------
// 抽象工厂接口
type ComputerFactory interface {
	CreateCpu() Cpu
	CreateMemory() Memory
}

// --------------------------------------------------------
// 具体工厂(computer1)的实现
type Computer1Factory struct{}

func (c *Computer1Factory) CreateCpu() Cpu {
	return &Cpu1{
		CpuBase: &CpuBase{},
	}
}
func (c *Computer1Factory) CreateMemory() Memory {
	return &Memory1{
		MemoryBase: &MemoryBase{},
	}
}

// 具体工厂(computer2)的实现
type Computer2Factory struct{}

func (c *Computer2Factory) CreateCpu() Cpu {
	return &Cpu2{
		CpuBase: &CpuBase{},
	}
}
func (c *Computer2Factory) CreateMemory() Memory {
	return &Memory2{
		MemoryBase: &MemoryBase{},
	}
}

// --------------------------------------------------------
// 抽象的产品接口
type Cpu interface {
	SetCpuBrand(b string)
	GetCpuBrand() string
}
type Memory interface {
	SetMemoryBrand(b string)
	GetMemoryBrand() string
}

// --------------------------------------------------------
// 具体产品的基结构体，封装公共方法
type CpuBase struct {
	cpuBrand string
}
type MemoryBase struct {
	memBrand string
}

func (c *CpuBase) SetCpuBrand(b string) {
	c.cpuBrand = b
}

func (c *CpuBase) GetCpuBrand() string {
	return c.cpuBrand
}

func (m *MemoryBase) SetMemoryBrand(b string) {
	m.memBrand = b
}

func (m *MemoryBase) GetMemoryBrand() string {
	return m.memBrand
}

// --------------------------------------------------------
// 具体产品结构体的实现
type Cpu1 struct {
	*CpuBase // 匿名成员
}
type Memory1 struct {
	*MemoryBase // 匿名成员
}

type Cpu2 struct {
	*CpuBase // 匿名成员
}
type Memory2 struct {
	*MemoryBase // 匿名成员
}

File: creational_patterns/cpp/abstract_factory/computer.cpp
/**
 * @file computer.cpp
 * @author tushanFox (tushanFox@outlook.com)
 * @brief 抽象工厂模式：一个抽象工厂可以生产一系列相关或相互依赖的抽象产品，这些抽象产品属于同一事物的不同子模块，而工厂方法模式只能生产一个抽象产品。
 *        比如一台电脑，有cpu/内存/显卡/硬盘等子模块，这些子模块具体对应不同的品牌、型号等等。
 * @version 0.1
 * @date 2022-01-12
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include "computer.h"
// 品牌1电脑的工厂
Cpu *Brand1ComputerFactory::CreateCpu() {
  return new Cpu1();
}
Memory *Brand1ComputerFactory::CreateMemory() {
  return new Memory1();
}

// 品牌2电脑的工厂
Cpu *Brand2ComputerFactory::CreateCpu() {
  return new Cpu2();
}
Memory *Brand2ComputerFactory::CreateMemory() {
  return new Memory2();
}


std::string Cpu1::GetCpuBrand(void) {
  return "Cpu Brand1";
}

std::string Memory1::GetMemoryBrand(void) {
  return "Memory Brand1";
}

std::string Cpu2::GetCpuBrand(void) {
  return "Cpu Brand2";
}

std::string Memory2::GetMemoryBrand(void) {
  return "Memory Brand2";
}